### PR TITLE
[SPARK-23268][SQL]Reorganize packages in data source V2

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?
1. create a new package for partitioning/distribution related classes.
    As Spark will add new concrete implementations of `Distribution` in new releases, it is good to 
    have a new package for partitioning/distribution related classes.

2. move streaming related class to package `org.apache.spark.sql.sources.v2.reader/writer.streaming`, instead of `org.apache.spark.sql.sources.v2.streaming.reader/writer`.
So that the there won't be package reader/writer inside package streaming, which is quite confusing.
Before change:
```
v2
├── reader
├── streaming
│   ├── reader
│   └── writer
└── writer
```

After change:
```
v2
├── reader
│   └── streaming
└── writer
    └── streaming
```
## How was this patch tested?
Unit test.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @jose-torres

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
cc @zsxwing @marmbrus too

**Comment by jose-torres** [comment]
Streaming part LGTM; I have no particular opinion or context on the distribution stuff.

**Comment by gengliangwang** [comment]
retest this please

**Comment by gatorsmile** [comment]
LGTM to adding the new package of partitioning/distribution.

**Comment by zsxwing** [comment]
cc @marmbrus

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceOffset.scala [hunk 1]
```diff
@@ -20,14 +20,16 @@ package org.apache.spark.sql.kafka010
 import org.apache.kafka.common.TopicPartition
 
 import org.apache.spark.sql.execution.streaming.{Offset, SerializedOffset}
-import org.apache.spark.sql.sources.v2.streaming.reader.{Offset => OffsetV2, PartitionOffset}
+import org.apache.spark.sql.sources.v2.reader.streaming
+import org.apache.spark.sql.sources.v2.reader.streaming.PartitionOffset
```
Comment (cloud-fan) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceOffset.scala]: can we keep it same as before?
```
import org.apache.spark.sql.sources.v2.reader.streaming.{Offset => OffsetV2, PartitionOffset}
```

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceOffset.scala [hunk 2]
```diff
@@ -20,14 +20,16 @@ package org.apache.spark.sql.kafka010
...
 
 /**
  * An [[Offset]] for the [[KafkaSource]]. This one tracks all partitions of subscribed topics and
```
Comment (ueshin) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceOffset.scala]: Should this `Offset` be `streaming.Offset`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/memoryV2.scala [hunk 3]
```diff
@@ -30,9 +30,8 @@ import org.apache.spark.sql.catalyst.plans.logical.{LeafNode, Statistics}
 import org.apache.spark.sql.catalyst.streaming.InternalOutputModes.{Append, Complete, Update}
 import org.apache.spark.sql.execution.streaming.Sink
 import org.apache.spark.sql.sources.v2.{DataSourceOptions, DataSourceV2}
-import org.apache.spark.sql.sources.v2.streaming.StreamWriteSupport
-import org.apache.spark.sql.sources.v2.streaming.writer.StreamWriter
-import org.apache.spark.sql.sources.v2.writer._
+import org.apache.spark.sql.sources.v2.writer.{StreamWriteSupport, _}
```
Comment (ueshin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/memoryV2.scala]: `import org.apache.spark.sql.sources.v2.writer._`?

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceOffset.scala [hunk 4]
```diff
@@ -20,14 +20,15 @@ package org.apache.spark.sql.kafka010
...
+import org.apache.spark.sql.sources.v2.reader.streaming.{Offset => OffsetV2, PartitionOffset}
 
 /**
  * An [[Offset]] for the [[KafkaSource]]. This one tracks all partitions of subscribed topics and
  * their offsets.
  */
 private[kafka010]
-case class KafkaSourceOffset(partitionToOffsets: Map[TopicPartition, Long]) extends OffsetV2 {
+case class KafkaSourceOffset(partitionToOffsets: Map[TopicPartition, Long])
+  extends OffsetV2 {
```
Comment (cloud-fan) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceOffset.scala]: unnecessary change?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala [hunk 5]
```diff
@@ -403,7 +403,7 @@ class MicroBatchExecution(
           val current = committedOffsets.get(reader).map(off => reader.deserializeOffset(off.json))
           reader.setOffsetRange(
             toJava(current),
-            Optional.of(available.asInstanceOf[OffsetV2]))
+            Optional.of(available.asInstanceOf[streaming.Offset]))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala]: shall we still use `OffsetV2`?

**Comment by cloud-fan** [comment]
LGTM, also  cc @rdblue

**Comment by gatorsmile** [comment]
Thanks! Merged to master/2.3
